Duplicati/Library/Backend/AzureBlob/AzureBlobWrapper.cs

```diff
@@ -112,46 +112,58 @@ public void DeleteObject(string keyName)
             _container.GetBlockBlobReference(keyName).DeleteIfExists();
         }
 
-        public virtual List<IFileEntry> ListContainerEntries()
+        public virtual IEnumerable<IFileEntry> ListContainerEntries(string prefix = null)
         {
-            var listBlobItems = _container.ListBlobs(blobListingDetails: BlobListingDetails.Metadata);
-            try
-            {
-                return listBlobItems.Select(x =>
-                {
-                    var absolutePath = x.StorageUri.PrimaryUri.AbsolutePath;
-                    var containerSegment = string.Concat("/", _containerName, "/");
-                    var blobName = absolutePath.Substring(absolutePath.IndexOf(
-                        containerSegment, System.StringComparison.Ordinal) + containerSegment.Length);
-
-                    try
-                    {
-                        if (x is CloudBlockBlob cb)
-                        {
-                            var lastModified = new System.DateTime();
-                            if (cb.Properties.LastModified != null)
-                                lastModified = new System.DateTime(cb.Properties.LastModified.Value.Ticks, System.DateTimeKind.Utc);
-                            return new FileEntry(Uri.UrlDecode(blobName.Replace("+", "%2B")), cb.Properties.Length, lastModified, lastModified);
-                        }
-                    }
-                    catch
-                    {
-                        // If the metadata fails to parse, return the basic entry
-                    }
-
-                    return new FileEntry(Uri.UrlDecode(blobName.Replace("+", "%2B")));
-                })
-                .Cast<IFileEntry>()
-                .ToList();
-            }
-            catch (StorageException ex)
-            {
-                if (ex.RequestInformation.HttpStatusCode == 404)
-                {
-                    throw new FolderMissingException(ex);
-                }
-                throw;
-            }
-        }
+            var listBlobItems = _container.ListBlobs(prefix: prefix, blobListingDetails: BlobListingDetails.Metadata);
+
+            var itemsEnumerator = listBlobItems.GetEnumerator();
+
+            while (true)
+            {
+                try
+                {
+                    if (!itemsEnumerator.MoveNext())
+                        break;
+                }
+                catch (StorageException ex)
+                when (ex.RequestInformation.HttpStatusCode == 404)
+                {
+                    itemsEnumerator.Dispose();
+                    throw new FolderMissingException(Strings.AzureBlobBackend.MissingContainerError(_containerName, ex.Message), ex);
+                }
+                catch
+                {
+                    itemsEnumerator.Dispose();
+                    throw;
+                }
+
+                yield return ParseBlobItem(itemsEnumerator.Current);
+            }

```

Comment:
This is the only way I could think of to return an `IEnumerable<>`. C# does not allow `yield return` from inside `try-catch` blocks.

Reply:
I don't have the ability to test this, but does a `using` statement allow us to remove the calls to `Dispose` as well as the final `catch` block ,while still disposing of everything correctly (in particular, with how we enumerate the items in the `AzureBlobBackend.Test` method)?
```
using (IEnumerator<IListBlobItem> itemsEnumerator = listBlobItems.GetEnumerator())
{
    while (true)
    {   
        try 
        {   
            if (!itemsEnumerator.MoveNext())
                break;
        }   
        catch (StorageException ex) when (ex.RequestInformation.HttpStatusCode == 404)
        {   
            throw new FolderMissingException(Strings.AzureBlobBackend.MissingContainerError(_containerName, ex.Message), ex);
        }   

        yield return ParseBlobItem(itemsEnumerator.Current);
    }   
}
```